File: cosky-discovery/src/main/java/me/ahoo/cosky/discovery/redis/RedisServiceRegistry.java
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.ahoo.cosky.discovery.redis;

import me.ahoo.cosky.discovery.InstanceIdGenerator;
import me.ahoo.cosky.discovery.NamespacedInstanceId;
import me.ahoo.cosky.discovery.RegistryProperties;
import me.ahoo.cosky.discovery.ServiceInstance;
import me.ahoo.cosky.discovery.ServiceInstanceCodec;
import me.ahoo.cosky.discovery.ServiceRegistry;

import com.google.common.base.Preconditions;
import com.google.common.base.Strings;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.core.ReactiveStringRedisTemplate;
import reactor.core.publisher.Mono;

import java.util.Arrays;
import java.util.Collections;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Redis Service Registry.
 *
 * @author ahoo wang
 */
@Slf4j
public class RedisServiceRegistry implements ServiceRegistry {
    
    private final RegistryProperties registryProperties;
    private final ReactiveStringRedisTemplate redisTemplate;
    private final ConcurrentHashMap<NamespacedInstanceId, ServiceInstance> registeredEphemeralInstances;
    
    public RedisServiceRegistry(RegistryProperties registryProperties,
                                ReactiveStringRedisTemplate redisTemplate) {
        this.registeredEphemeralInstances = new ConcurrentHashMap<>();
        this.registryProperties = registryProperties;
        this.redisTemplate = redisTemplate;
    }
    
    private Mono<Boolean> register0(String namespace, ServiceInstance serviceInstance) {
        /**
         * ARGV
         */
        String[] infoArgs = {
            /**
             * local instanceTtl = ARGV[1];
            */
            serviceInstance.isEphemeral() ? String.valueOf(registryProperties.getInstanceTtl()) : "-1",
            /**
             * local serviceId = ARGV[2];
            */
            serviceInstance.getServiceId(),
            /**
             * local instanceId = ARGV[3];
            */
            serviceInstance.getInstanceId(),
            /**
             * local scheme = ARGV[4];
            */
            serviceInstance.getSchema(),
            /**
             * local host = ARGV[5];
            */
            serviceInstance.getHost(),
            /**
             * local port = ARGV[6];
            */
            String.valueOf(serviceInstance.getPort()),
            /**
             * local weight = ARGV[7];
            */
            String.valueOf(serviceInstance.getWeight())
        };
        
        
        String[] values = ServiceInstanceCodec.encodeMetadata(infoArgs, serviceInstance.getMetadata());
        
        return redisTemplate.execute(
                DiscoveryRedisScripts.SCRIPT_REGISTRY_REGISTER,
                Collections.singletonList(namespace),
                Arrays.asList(values)
            )
            .next();
    }
    
    @Override
    public Mono<Boolean> setService(String namespace, String serviceId) {
        Preconditions.checkArgument(!Strings.isNullOrEmpty(namespace), "namespace can not be empty!");
        Preconditions.checkArgument(!Strings.isNullOrEmpty(serviceId), "serviceId can not be empty!");
        
        if (log.isInfoEnabled()) {
            log.info("setService - serviceId:[{}]  @ namespace:[{}].", serviceId, namespace);
        }
        return redisTemplate.execute(
                DiscoveryRedisScripts.SCRIPT_REGISTRY_SET_SERVICE,
                Collections.singletonList(namespace),
                Collections.singletonList(serviceId)
            )
            .next();
    }
    
    @Override
    public Mono<Boolean> removeService(String namespace, String serviceId) {
        Preconditions.checkArgument(!Strings.isNullOrEmpty(namespace), "namespace can not be empty!");
        Preconditions.checkArgument(!Strings.isNullOrEmpty(serviceId), "serviceId can not be empty!");
        
        if (log.isWarnEnabled()) {
            log.warn("removeService - serviceId:[{}]  @ namespace:[{}].", serviceId, namespace);
        }
        return redisTemplate.execute(
                DiscoveryRedisScripts.SCRIPT_REGISTRY_REMOVE_SERVICE,
                Collections.singletonList(namespace),
                Collections.singletonList(serviceId)
            )
            .next();
    }
    
    @Override
    public Mono<Boolean> register(String namespace, ServiceInstance serviceInstance) {
        ensureNamespacedInstance(namespace, serviceInstance);
        
        ensureInstanceId(serviceInstance);
        if (log.isInfoEnabled()) {
            log.info("register - instanceId:[{}]  @ namespace:[{}].", serviceInstance.getInstanceId(), namespace);
        }
        
        addEphemeralInstance(namespace, serviceInstance);
        return register0(namespace, serviceInstance);
    }
    
    private void ensureNamespacedInstance(String namespace, ServiceInstance serviceInstance) {
        Preconditions.checkArgument(!Strings.isNullOrEmpty(namespace), "namespace can not be empty!");
        Preconditions.checkNotNull(serviceInstance, "serviceInstance can not be null!");
        Preconditions.checkArgument(!Strings.isNullOrEmpty(serviceInstance.getServiceId()), "serviceId can not be empty!");
        Preconditions.checkArgument(!Strings.isNullOrEmpty(serviceInstance.getSchema()), "schema can not be empty!");
        Preconditions.checkArgument(!Strings.isNullOrEmpty(serviceInstance.getHost()), "host can not be empty!");
    }
    
    private void ensureInstanceId(ServiceInstance serviceInstance) {
        if (Strings.isNullOrEmpty(serviceInstance.getInstanceId())) {
            serviceInstance.setInstanceId(InstanceIdGenerator.DEFAULT.generate(serviceInstance));
        }
    }
    
    private void addEphemeralInstance(String namespace, ServiceInstance serviceInstance) {
        if (!serviceInstance.isEphemeral()) {
            return;
        }
        registeredEphemeralInstances.put(NamespacedInstanceId.of(namespace, serviceInstance.getInstanceId()), serviceInstance);
    }
    
    private void removeEphemeralInstance(String namespace, String instanceId) {
        registeredEphemeralInstances.remove(NamespacedInstanceId.of(namespace, instanceId));
    }
    
    private void removeEphemeralInstance(String namespace, ServiceInstance serviceInstance) {
        if (!serviceInstance.isEphemeral()) {
            return;
        }
        
        registeredEphemeralInstances.remove(NamespacedInstanceId.of(namespace, serviceInstance.getInstanceId()));
    }
    
    @Override
    public Map<NamespacedInstanceId, ServiceInstance> getRegisteredEphemeralInstances() {
        return registeredEphemeralInstances;
    }
    
    @Override
    public Mono<Boolean> setMetadata(String namespace, String serviceId, String instanceId, String key, String value) {
        String[] values = {instanceId, key, value};
        return setMetadata0(namespace, instanceId, values);
    }
    
    @Override
    public Mono<Boolean> setMetadata(String namespace, String serviceId, String instanceId, Map<String, String> metadata) {
        String[] values = ServiceInstanceCodec.encodeMetadata(new String[] {instanceId}, metadata);
        return setMetadata0(namespace, instanceId, values);
    }
    
    private Mono<Boolean> setMetadata0(String namespace, String instanceId, String[] args) {
        Preconditions.checkArgument(!Strings.isNullOrEmpty(namespace), "namespace can not be empty!");
        Preconditions.checkArgument(!Strings.isNullOrEmpty(instanceId), "instanceId can not be empty!");
        
        if (log.isInfoEnabled()) {
            log.info("setMetadata - instanceId:[{}] @ namespace:[{}].", instanceId, namespace);
        }
        return redisTemplate.execute(
                DiscoveryRedisScripts.SCRIPT_REGISTRY_SET_METADATA,
                Collections.singletonList(namespace),
                Arrays.asList(args)
            )
            .next();
    }
    
    @Override
    public Mono<Boolean> renew(String namespace, ServiceInstance serviceInstance) {
        ensureNamespacedInstance(namespace, serviceInstance);
        Preconditions.checkArgument(!Strings.isNullOrEmpty(serviceInstance.getInstanceId()), "instanceId can not be empty!");
        
        if (log.isDebugEnabled()) {
            log.debug("renew - instanceId:[{}] @ namespace:[{}].", serviceInstance.getInstanceId(), namespace);
        }
        
        if (!serviceInstance.isEphemeral()) {
            if (log.isWarnEnabled()) {
                log.warn("renew - instanceId:[{}] @ namespace:[{}] is not ephemeral, can not renew.", serviceInstance.getInstanceId(), namespace);
            }
            return Mono.just(Boolean.FALSE);
        }

        return redisTemplate.execute(
                DiscoveryRedisScripts.SCRIPT_REGISTRY_RENEW,
                Collections.singletonList(namespace),
                Arrays.asList(serviceInstance.getInstanceId(), String.valueOf(registryProperties.getInstanceTtl()))
            )
            .flatMap(status -> {
                if (status <= 0) {
                    if (log.isWarnEnabled()) {
                        log.warn("renew - instanceId:[{}] @ namespace:[{}] status is [{}],register again.", serviceInstance.getInstanceId(), namespace, status);
                    }
                    return register(namespace, serviceInstance);
                }
                return Mono.just(Boolean.TRUE);
            })
            .next();
    }
    
    @Override
    public Mono<Boolean> deregister(String namespace, String serviceId, String instanceId) {
        if (log.isInfoEnabled()) {
            log.info("deregister - instanceId:[{}] @ namespace:[{}].", instanceId, namespace);
        }
        removeEphemeralInstance(namespace, instanceId);
        
        return deregister0(namespace, serviceId, instanceId);
    }
    
    @Override
    public Mono<Boolean> deregister(String namespace, ServiceInstance serviceInstance) {
        ensureInstanceId(serviceInstance);
        if (log.isInfoEnabled()) {
            log.info("deregister - instanceId:[{}] @ namespace:[{}].", serviceInstance.getInstanceId(), namespace);
        }
        
        removeEphemeralInstance(namespace, serviceInstance);
        return deregister0(namespace, serviceInstance.getServiceId(), serviceInstance.getInstanceId());
    }
    
    private Mono<Boolean> deregister0(String namespace, String serviceId, String instanceId) {
        Preconditions.checkArgument(!Strings.isNullOrEmpty(namespace), "namespace can not be empty!");
        Preconditions.checkArgument(!Strings.isNullOrEmpty(serviceId), "serviceId can not be empty!");
        Preconditions.checkArgument(!Strings.isNullOrEmpty(instanceId), "instanceId can not be empty!");
        return redisTemplate.execute(
                DiscoveryRedisScripts.SCRIPT_REGISTRY_DEREGISTER,
                Collections.singletonList(namespace),
                Arrays.asList(serviceId, instanceId)
            )
            .next();
    }
    
}

File: cosky-discovery/src/main/kotlin/me/ahoo/cosky/discovery/redis/RedisServiceRegistry.kt
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.ahoo.cosky.discovery.redis

import io.github.oshai.kotlinlogging.KotlinLogging
import me.ahoo.cosky.discovery.NamespacedInstanceId
import me.ahoo.cosky.discovery.RegistryProperties
import me.ahoo.cosky.discovery.ServiceInstance
import me.ahoo.cosky.discovery.ServiceInstanceCodec.encodeMetadata
import me.ahoo.cosky.discovery.ServiceInstanceCodec.encodeMetadataKey
import me.ahoo.cosky.discovery.ServiceRegistry
import org.springframework.data.redis.core.ReactiveStringRedisTemplate
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toMono
import java.util.concurrent.ConcurrentHashMap

/**
 * Redis Service Registry.
 *
 * @author ahoo wang
 */
class RedisServiceRegistry(
    private val registryProperties: RegistryProperties,
    private val redisTemplate: ReactiveStringRedisTemplate
) : ServiceRegistry {
    companion object {
        private val log = KotlinLogging.logger {}
    }

    override val registeredEphemeralInstances: ConcurrentHashMap<NamespacedInstanceId, ServiceInstance> =
        ConcurrentHashMap()

    private fun registerInternal(namespace: String, serviceInstance: ServiceInstance): Mono<Boolean> {
        val argVCapacity = 6 + serviceInstance.metadata.size * 2

        /**
         * ARGV
         */
        val values = buildList(argVCapacity) {
            add(if (serviceInstance.isEphemeral) registryProperties.instanceTtl.seconds.toString() else "-1")
            add(serviceInstance.serviceId)
            add(serviceInstance.instanceId)
            add(serviceInstance.schema)
            add(serviceInstance.host)
            add(serviceInstance.port.toString())
            add(serviceInstance.weight.toString())
            encodeMetadata(this, serviceInstance.metadata)
        }
        return redisTemplate.execute(
            DiscoveryRedisScripts.SCRIPT_REGISTRY_REGISTER,
            listOf(namespace),
            values,
        ).next()
    }

    override fun setService(namespace: String, serviceId: String): Mono<Boolean> {
        require(namespace.isNotBlank()) { "namespace must not be blank!" }
        require(serviceId.isNotBlank()) { "serviceId must not be blank!" }
        log.info {
            "Set Service - serviceId:[$serviceId]  @ namespace:[$namespace]."
        }
        return redisTemplate.execute(
            DiscoveryRedisScripts.SCRIPT_REGISTRY_SET_SERVICE,
            listOf(namespace),
            listOf(serviceId),
        ).next()
    }

    override fun removeService(namespace: String, serviceId: String): Mono<Boolean> {
        require(namespace.isNotBlank()) { "namespace must not be blank!" }
        require(serviceId.isNotBlank()) { "serviceId must not be blank!" }
        log.warn {
            "Remove Service - serviceId:[$serviceId]  @ namespace:[$namespace]."
        }
        return redisTemplate.execute(
            DiscoveryRedisScripts.SCRIPT_REGISTRY_REMOVE_SERVICE,
            listOf(namespace),
            listOf(serviceId),
        ).next()
    }

    override fun register(namespace: String, serviceInstance: ServiceInstance): Mono<Boolean> {
        require(namespace.isNotBlank()) { "namespace must not be blank!" }
        log.info {
            "Register - instanceId:[${serviceInstance.instanceId}]  @ namespace:[$namespace]."
        }
        return registerInternal(namespace, serviceInstance).doOnSubscribe {
            addEphemeralInstance(namespace, serviceInstance)
        }
    }

    private fun addEphemeralInstance(namespace: String, serviceInstance: ServiceInstance) {
        if (!serviceInstance.isEphemeral) {
            return
        }
        registeredEphemeralInstances[NamespacedInstanceId(namespace, serviceInstance.instanceId)] =
            serviceInstance
    }

    private fun removeEphemeralInstance(namespace: String, instanceId: String) {
        registeredEphemeralInstances.remove(NamespacedInstanceId(namespace, instanceId))
    }

    private fun removeEphemeralInstance(namespace: String, serviceInstance: ServiceInstance) {
        if (!serviceInstance.isEphemeral) {
            return
        }
        registeredEphemeralInstances.remove(NamespacedInstanceId(namespace, serviceInstance.instanceId))
    }

    override fun setMetadata(
        namespace: String,
        serviceId: String,
        instanceId: String,
        key: String,
        value: String
    ): Mono<Boolean> {
        val values = listOf(instanceId, encodeMetadataKey(key), value)
        return setMetadataInternal(namespace, instanceId, values)
    }

    override fun setMetadata(
        namespace: String,
        serviceId: String,
        instanceId: String,
        metadata: Map<String, String>
    ): Mono<Boolean> {
        val argVCapacity = 1 + metadata.size * 2
        val values = buildList(argVCapacity) {
            add(instanceId)
            encodeMetadata(this, metadata)
        }
        return setMetadataInternal(namespace, instanceId, values)
    }

    private fun setMetadataInternal(namespace: String, instanceId: String, args: List<String>): Mono<Boolean> {
        require(namespace.isNotBlank()) { "namespace must not be blank!" }
        require(instanceId.isNotBlank()) { "instanceId must not be blank!" }
        log.info {
            "Set Metadata - instanceId:[$instanceId] @ namespace:[$namespace]."
        }
        return redisTemplate.execute(
            DiscoveryRedisScripts.SCRIPT_REGISTRY_SET_METADATA,
            listOf(namespace),
            args,
        )
            .next()
    }

    override fun renew(namespace: String, serviceInstance: ServiceInstance): Mono<Boolean> {
        require(namespace.isNotBlank()) { "namespace must not be blank!" }
        log.debug {
            "Renew - instanceId:[${serviceInstance.instanceId}] @ namespace:[$namespace]."
        }
        if (!serviceInstance.isEphemeral) {
            log.warn {
                "Renew - instanceId:[${serviceInstance.instanceId}] @ namespace:[$namespace] is not ephemeral, can not renew."
            }
            return false.toMono()
        }
        return redisTemplate.execute(
            DiscoveryRedisScripts.SCRIPT_REGISTRY_RENEW,
            listOf(namespace),
            listOf(serviceInstance.instanceId, registryProperties.instanceTtl.seconds.toString()),
        )
            .flatMap { status ->
                if (status <= 0) {
                    log.warn {
                        "Renew - instanceId:[${serviceInstance.instanceId}] @ namespace:[$namespace] status is [$status],register again."
                    }
                    return@flatMap register(namespace, serviceInstance)
                }
                true.toMono()
            }
            .next()
    }

    override fun deregister(namespace: String, serviceId: String, instanceId: String): Mono<Boolean> {
        log.info {
            "Deregister - instanceId:[$instanceId] @ namespace:[$namespace]."
        }
        return deregisterInternal(namespace, serviceId, instanceId)
            .doOnSubscribe {
                removeEphemeralInstance(namespace, instanceId)
            }
    }

    override fun deregister(namespace: String, serviceInstance: ServiceInstance): Mono<Boolean> {
        log.info {
            "Deregister - instanceId:[${serviceInstance.instanceId}] @ namespace:[$namespace]."
        }
        return deregisterInternal(namespace, serviceInstance.serviceId, serviceInstance.instanceId)
            .doOnSubscribe {
                removeEphemeralInstance(namespace, serviceInstance)
            }
    }

    private fun deregisterInternal(namespace: String, serviceId: String, instanceId: String): Mono<Boolean> {
        require(namespace.isNotBlank()) { "namespace must not be blank!" }
        require(serviceId.isNotBlank()) { "serviceId must not be blank!" }
        require(instanceId.isNotBlank()) { "instanceId must not be blank!" }
        return redisTemplate.execute(
            DiscoveryRedisScripts.SCRIPT_REGISTRY_DEREGISTER,
            listOf(namespace),
            listOf(serviceId, instanceId),
        ).next()
    }
}
